Panic with plan -out=tfplan -destroy -refresh=false

### Terraform Version

```
Terraform v0.15.0-rc1
on darwin_amd64
+ provider registry.terraform.io/hashicorp/null v3.1.0
```

### Terraform Configuration Files

```terraform
resource "null_resource" "none" {}
```

### Debug Output

### Crash Output

```
panic: attempt to write nil state to file

goroutine 51 [running]:
github.com/hashicorp/terraform/states/statefile.writeStateV4(0xc000c04b10, 0x37d0d80, 0xc000d43360, 0xc000986a50, 0xc000d43360, 0x3841de0)
	/home/circleci/project/project/states/statefile/version4.go:309 +0x19f4
github.com/hashicorp/terraform/states/statefile.Write(0xc000c04b10, 0x37d0d80, 0xc000d43360, 0xc000d43360, 0x0)
	/home/circleci/project/project/states/statefile/write.go:15 +0x5c
github.com/hashicorp/terraform/plans/planfile.Create(0x7ffeefbff8ed, 0x6, 0xc00000f6f0, 0xc000c04b10, 0xc0002d0f50, 0x0, 0x0)
	/home/circleci/project/project/plans/planfile/writer.go:57 +0x57a
github.com/hashicorp/terraform/backend/local.(*Local).opPlan(0xc0004beaa0, 0x382b398, 0xc0003f5640, 0x382b398, 0xc0003f5680, 0xc000c74700, 0xc0003f5600)
	/home/circleci/project/project/backend/local/backend_plan.go:121 +0x685
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc00078db30, 0xc00078db40, 0xc00078db50, 0xc0004beaa0, 0xc00078db20, 0x382b398, 0xc0003f5640, 0x382b398, 0xc0003f5680, 0xc000c74700, ...)
	/home/circleci/project/project/backend/local/backend.go:321 +0xe8
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
	/home/circleci/project/project/backend/local/backend.go:315 +0x2f2
```

### Expected Behavior

Terraform works

### Actual Behavior

Terraform does not work!!!

### Steps to Reproduce

1. `terraform init`
2. `terraform apply -auto-approve`
3. `terraform plan -refresh=false -out=tfplan -destroy`

### Additional Context

### References

Found while trying to reproduce #28266

--- a/terraform/context.go
+++ b/terraform/context.go
@@ -604,7 +604,9 @@ func (c *Context) plan() (*plans.Plan, tfdiags.Diagnostics) {
 
 func (c *Context) destroyPlan() (*plans.Plan, tfdiags.Diagnostics) {
 	var diags tfdiags.Diagnostics
-	destroyPlan := &plans.Plan{}
+	destroyPlan := &plans.Plan{
+		State: c.state.DeepCopy(),
+	}
 	c.changes = plans.NewChanges()
 
 	// A destroy plan starts by running Refresh to read any pending data

--- a/terraform/context_plan2_test.go
+++ b/terraform/context_plan2_test.go
@@ -333,3 +333,47 @@ resource "test_resource" "b" {
 		}
 	}
 }
+
+func TestContext2Plan_destroySkipRefresh(t *testing.T) {
+	m := testModuleInline(t, map[string]string{
+		"main.tf": `
+resource "test_object" "a" {
+}
+`,
+	})
+
+	p := simpleMockProvider()
+
+	addr := mustResourceInstanceAddr("test_object.a")
+	state := states.BuildState(func(s *states.SyncState) {
+		s.SetResourceInstanceCurrent(addr, &states.ResourceInstanceObjectSrc{
+			AttrsJSON: []byte(`{"test_string":"foo"}`),
+			Status:    states.ObjectReady,
+		}, mustProviderConfig(`provider["registry.terraform.io/hashicorp/test"]`))
+	})
+
+	ctx := testContext2(t, &ContextOpts{
+		Config: m,
+		State:  state,
+		Providers: map[addrs.Provider]providers.Factory{
+			addrs.NewDefaultProvider("test"): testProviderFuncFixed(p),
+		},
+		Destroy:     true,
+		SkipRefresh: true,
+	})
+
+	plan, diags := ctx.Plan()
+	if diags.HasErrors() {
+		t.Fatal(diags.Err())
+	}
+
+	if plan.State == nil {
+		t.Fatal("missing plan state")
+	}
+
+	for _, c := range plan.Changes.Resources {
+		if c.Action != plans.Delete {
+			t.Errorf("unexpected %s change for %s", c.Action, c.Addr)
+		}
+	}
+}

